### PR TITLE
Survival Mode first iteration

### LINKED ISSUE #31: hardcore mode
- map resets when everyone dies
- disable respawning and only respawn on checkpoints
- put dead players into spectate

### PR DESCRIPTION
**Pull request by Balimbanana**
Fixes #31 
This will add Survival Mode and a method to revive players by the additional plugin.
- Added CVar sourcecoop_survivalmode default 0.
1 will respawn all players once the last player has died.
2 will restart the map when all players have died.
Players will respawn at any `trigger_autosave` or any `Save` or `SaveDangerous` (with restrictions) to `logic_autosave`.
- Added CBaseEntity::GetGroundEntity() currently not used in SrcCoop plugins, but it is a good thing to have for the future.
- Fix to CBaseEntity::IsAlive() to use `m_lifeState` instead of health, as dead entities can on occasion actually have +0 health. This is how it is checked in the SDK as well.
```
// m_lifeState values
#define	LIFE_ALIVE				0 // alive
#define	LIFE_DYING				1 // playing death animation or still falling off of a ledge waiting to hit ground
#define	LIFE_DEAD				2 // dead. lying still.
#define LIFE_RESPAWNABLE		3
#define LIFE_DISCARDBODY		4
```

Added SrcCoop_Addon_Revive
- Enables reviving players by holding E near the dead players body.
- Console command: srccoop_revive takes argument of client index, this may be extended to allow group selection or player name, or !self
- ConVar sourcecoop_revive_time default 4.0 sets the amount of time you have to hold down E to revive a player. Longer values might need some sort of progress indicator because there isn't one at the moment.
- ConVar sourcecoop_revive_score default 1, gives this amount of score to the reviver.

### CODE REVIEW

**Comment by Alienmario** [comment]
Hi, is it ready for review?
Can I add commits to this PR? I would reorganize some things and add polish where needed.

**Comment by Balimbanana** [comment]
Alright.
I was considering changing the CanSpawn check to use a global bool and resetting it on disconnect per client, but I had found some unused properties on the client such as the m_iRespawnFrames which is reset automatically on player entity creation.

**Comment by BlueRaja** [comment]
If you can provide compiled binaries we'd be happy to help test.

I tried compiling them myself, but there were compile errors in files you didn't touch.  After fixing them _(eg. having to `#include <regex>` in utils.inc; and change `CEntityOutputLump[1] pOther` to `CEntityOutputLump pOther[1]` in levellump.inc)_ and running the server, everyone spawned in the ceiling and coop wasn't working.  So clearly I'm doing something wrong.

**Comment by Balimbanana** [comment]
It will compile if it is done on the same version as in the readme 6645 as per issues in later versions: https://github.com/ampreeT/SourceCoop/issues/58#issuecomment-1004892233

**Comment by Balimbanana** [comment]
1. The revive plugin must be compiled at the same version and same includes as SrcCoop, as I found that while it doesn't give any errors for forwards/natives not existing, they just don't work at all. Calling `pPlayer->SetCanSpawn(true);` was doing nothing when it was compiled on a different version.

2. It is intended, although I should add a message of some kind, designed to revive all players at the games checkpoints like `trigger_autosave` and `Save` inputs to `logic_relay` I could add a CVar to disable this functionality as well if you don't want it. But some maps are a bit long if someone falls off an edge and you couldn't get to their body it would be a while before they can get back in.

3. I will have to look in to what checkpoints and systems are not cleared on map start, it appears to be keeping some data over a map change.

4. Which maps and sections did this happen on in particular? It might be the hook for `player_loadsaved` causing a `FFADE_STAYOUT` for the !activator. I could also potentially just add it in to the spawn function to do a `FFADE_PURGE`.

5. Looks like an issue in https://github.com/ampreeT/SourceCoop/blob/master/scripting/include/srccoop/manager.inc#L415 the second bool is for `bAliveOnly` meaning if it is false, it counts dead players. I am not sure if there is a particular reason it should be counting dead players in the map vote?

**Comment by BlueRaja** [comment]
> Which maps and sections did this happen on in particular?

There is a map after `c2a5a` where you start right next to a cliff and can accidentally fall off.  I think it was `c2a5d`??  If you do that, it fades to black indefinitely.

Also a friend fell from a tower onto normal ground and died in `c2a5f`, resulting in that indefinite black screen.

> The revive plugin must be compiled at the same version and same includes as SrcCoop

Yes, I compiled all the `srccoop*.sp` files from this branch at the same time using the same version of SourceMod.

**Comment by Balimbanana** [comment]
I found that there is actually another issue to do with the bools being separated per plugin, while players are set to be able to spawn in the revive plugin, it is not actually applying to the srccoop plugin which is causing the conflict. It was working before when I was using the unused player properties because those are stored on the player entity separate from any one plugin holding their own declarations.

**Comment by Balimbanana** [comment]
Yeah, it looks like `player_loadsaved` is being called there with the !activator of a player upon falling.

bm_c2a5c
```
{
"model" "*60"
"wait" "1"
"targetname" "fall_off"
"StartDisabled" "0"
"spawnflags" "1"
"origin" "-9491 -8611.5 5219.5"
"classname" "trigger_multiple"
"hammerid" "555097"
"OnStartTouch" "reload,Reload,,0,-1"
}
```
bm_c2a5f has a couple occasions involving the death of the guard and another trigger elsewhere that could cause a reload input.

**Comment by Balimbanana** [comment]
I wasn't able to reproduce the incorrect checkpoint being set, is there a particular case where it always happens?

**Comment by BlueRaja** [comment]
If I remember correctly, beating `bm_c2a4f` leads to `bm_c2a4g`, and beating that leads back to `bm_c2a4f` but at a different spawn point.  If you then die, the map restarts at the first checkpoint.

The same thing happens earlier with the tentacle monster, where you need to take a detour and come back later to unlock the chamber.  Something like `bm_c1a4b` -> `bm_c1a4c` -> bm_c1a4d` -> `bm_c1a4d`

**Comment by Balimbanana** [comment]
Ah, those ones will need some additional transition/global states detection.

**Comment by Balimbanana** [comment]
It is now in, tested c1a4b - c1a4d - c1a4b `kill` restarting the map. Then tested from c1a4b - c1a4c - c1a4b `kill` restarting map, and all globals are still active and entry point is correct.

**Comment by Alienmario** [comment]
> It is intended, although I should add a message of some kind, designed to revive all players at the games checkpoints like trigger_autosave and Save inputs to logic_relay I could add a CVar to disable this functionality as well if you don't want it.

We need to be able to fine control it per map. Adding a global cvar to disable respawning completely is good too.

>  But some maps are a bit long if someone falls off an edge and you couldn't get to their body it would be a while before they can get back in.

Better make sure to not fall then :)
Also, make sure players can't respawn by just rejoining.

> Some of the levels start over at the wrong place

All you have to do is add the reason from ForceChangeLevel to manager/ShouldResetGlobalStates (_or use SM_NORMAL_MAPCHANGE in ForceChangeLevel, but former is preferred_)

> Looks like an issue in https://github.com/ampreeT/SourceCoop/blob/master/scripting/include/srccoop/manager.inc#L415 the second bool is for bAliveOnly meaning if it is false, it counts dead players. I am not sure if there is a particular reason it should be counting dead players in the map vote?

Changing this is likely fine, just gonna have to deal with variations in the tick speed as players die/respawn. Only other problem I see is with 2 players, one player trying to catchup, dying and map changing instantly (?)

**Comment by Balimbanana** [comment]
I can do the rejoin block and CVar for blocking spawns by checkpoints in a bit, I have to go for now though.

**Comment by Balimbanana** [comment]
The rejoin block and full prevention of respawns by CVar `sourcecoop_disable_respawn`
I also changed the changelevel countdown in manager to check survival mode or disabled respawns for the count.
That way when playing in normal modes the counter won't be changing ticks constantly as players die and respawn.

File: scripting/include/srccoop/checkpoint.inc [hunk 1]
```diff
@@ -685,6 +685,33 @@ enum struct CCoopSpawn
...
+		// Could maybe add a translation for Checkpoint, although most people won't pay too much attention
+		SetHudTextParams(-1.0, 0.9, 1.0, 200, 200, 200, 255, 0, 1.0, 1.0, 1.0);
+		for (int i = 1; i <= MaxClients; i++)
+		{
+			CBasePlayer pPlayer = CBasePlayer(i);
+			if (pPlayer.IsValid())
+			{
+				if (!pPlayer.IsAlive())
+				{
+					SetPlayerCanSpawn(pPlayer);
```
Comment (BlueRaja) [inline scripting/include/srccoop/checkpoint.inc:706]:
```suggestion
					PlayerSetCanSpawn(pPlayer);
```

File: scripting/include/srccoop/playerpatch.inc [hunk 2]
```diff
@@ -177,6 +181,40 @@ public MRESReturn Hook_SetSuitUpdatePost(int _this, Handle hParams)
...
+				// this is mostly being caused by player_loadsaved having FFADE_STAYOUT but it could come from other sources as well
+				Client_ScreenFade(pPlayer.GetEntIndex(), 1, FFADE_IN|FFADE_PURGE, 1, 0, 0, 0, 0);
+				
+				return MRES_Supercede;
+			}
+			
+			// This is fired upon joining the game before selecting a team so this must be checked
+			if (pPlayer.GetTeam() > 1)
+			{
+				SetPlayerCanSpawn(pPlayer, false);
```
Comment (BlueRaja) [inline scripting/include/srccoop/playerpatch.inc:211]:
```suggestion
				PlayerSetCanSpawn(pPlayer, false);
```

File: scripting/srccoop.sp [hunk 3]
```diff
@@ -617,6 +637,111 @@ public Action Event_BroadcastTeamsound(Event hEvent, const char[] szName, bool b
...
+						}
+					}
+				}
+				else
+				{
+					CBasePlayer pPlayer = CBasePlayer(iVictimCheck);
+				
+					if (pPlayer.IsValid())
+					{
+						SetPlayerCanSpawn(pPlayer, false);
```
Comment (BlueRaja) [inline scripting/srccoop.sp:694]:
```suggestion
						PlayerSetCanSpawn(pPlayer, false);
```

File: scripting/srccoop.sp [hunk 4]
```diff
@@ -741,3 +866,19 @@ public Action Command_DumpMapEntities(int iArgs)
...
+{
+	if ((pConvar == g_pConvarPreventRespawn) || ((pConvar == g_pConvarSurvivalMode) && (!g_pConvarPreventRespawn.BoolValue)))
+	{
+		if (!pConvar.BoolValue)
+		{
+			g_pDeadPlayerIDs.Clear();
+			for (int i = 1;i <= MaxClients; i++)
+			{
+				CBasePlayer pPlayer = CBasePlayer(i);
+				SetPlayerCanSpawn(pPlayer);
```
Comment (BlueRaja) [inline scripting/srccoop.sp:880]:
```suggestion
				PlayerSetCanSpawn(pPlayer);
```

File: scripting/srccoop_addon_revive.sp [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+		return Plugin_Handled;
+	}
+	char szTargIndex[4];
+	GetCmdArg(1, szTargIndex, sizeof(szTargIndex));
+	CBasePlayer pTarget = CBasePlayer(StringToInt(szTargIndex));
+	if (pTarget.IsValid())
+	{
+		PrintToConsole(client, "Spawned %i", pTarget.GetEntIndex());
+		
+		SetPlayerCanSpawn(pTarget, true);
```
Comment (BlueRaja) [inline scripting/srccoop_addon_revive.sp:54]:
```suggestion
		PlayerSetCanSpawn(pTarget, true);
```

File: scripting/srccoop_addon_revive.sp [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+							StopSound(client, SNDCHAN_STATIC, "items/suitchargeok1.wav");
+							EmitSoundToAll("items/suitchargeno1.wav", client, SNDCHAN_AUTO, SNDLEVEL_NORMAL);
+							
+							return Plugin_Continue;
+						}
+						if (GetGameTime() >= g_flReviveTime[client])
+						{
+							EmitSoundToAll("weapons/tau/gauss_undercharge.wav", client, SNDCHAN_AUTO, SNDLEVEL_NORMAL, _, _, 150);
+							
+							SetPlayerCanSpawn(g_pReviveTarget[client], true);
```
Comment (BlueRaja) [inline scripting/srccoop_addon_revive.sp:107]:
```suggestion
							PlayerSetCanSpawn(g_pReviveTarget[client], true);
```

**Comment by Balimbanana** [comment]
Ah, I changed all those along with the native name in the latest.

**Comment by Alienmario** [comment]
I'm not sure why SteamIDs are needed at all, if you dont let any player spawn when map's in progress. If you don't wanna mess with the team-change hook, I can do it later as part of my review.

**Comment by BlueRaja** [comment]
The newest version works really well! We did find two bugs with revive:

* Players who are revived lose their HUD, and can't turn on their flashlight or switch weapons
* The revive trigger is invisible, and is not necessarily in the same place that their body is. We've had to run around like chickens with their heads cut off, spamming `e` to find the trigger.
  Some sort of visual indicator (literally anything) would be immensely helpful.

**Comment by Balimbanana** [comment]
I think using SteamID's would be good because newly joining players don't potentially get stuck in spectate until the end of the map, it will just put people off joining any survival servers if they don't join at the right time.

The HUD on revive sounds like the suit is not being applied, I'll have to check in to that.
The reason it doesn't line up is because it is using the server-side ragdoll, but client-side creates its own ragdolls. I could try and find a suitable indicator or sprite or something.

**Comment by Alienmario** [comment]
> I think using SteamID's would be good because newly joining players don't potentially get stuck in spectate until the end of the map, it will just put people off joining any survival servers if they don't join at the right time.

I guess that's true for BM in its current state. It is however a good candidate for a convar. Also, let's prefix survival convars with sourcecoop_survival_.

> The HUD on revive sounds like the suit is not being applied, I'll have to check in to that.

Suit is currently being reset in Hook_PlayerSpawnPost.

> The reason it doesn't line up is because it is using the server-side ragdoll, but client-side creates its own ragdolls. I could try and find a suitable indicator or sprite or something.

Before that, we should try find if there is a way to force serverside ragdoll. Either through commands or code.

**Comment by Balimbanana** [comment]
I will change the prefix for the survival convars.

There are things like `ai_force_serverside_ragdolls` which does make player ragdolls server-side, but it can also cause crashes in a lot of scenarios, or place players in a nullified state if their ragdoll is deleted. Could also catch the creation of `player_ragdoll` and replace it with something else, or create a dummy sort of ragdoll at the same origin. There is also the property of `m_bForceServerRagdoll` but it doesn't seem to make a difference.

I believe the issue with the suit not applying is from teleporting the player immediately after spawn, before spawnpost, I have changed the addon plugin to wait 0.1 seconds to teleport which *appears* to work for equipping the player. You can see it by going to the active checkpoint, there is usually a spray of weapons that are all removed right after each spawn which is not being given to the revived player.

I was able to use this as a marker, it is just the health icon as a sprite and can change the color to whichever if need be.
image

**Comment by BlueRaja** [comment]
Sorry for the late response; it takes a while to get everyone together.

With the latest version, `sourcecoop_survivalmode` is no longer working.  We have it set to `2`, but people are still reviving instantly.

**Comment by BlueRaja** [comment]
Unfortunately the new health symbols do not accurately show where the revive location is :(

**Comment by Alienmario** [comment]
Balim, did you mean to close it?

**Pull request closed without merging**